File: 189. Rotate Array.cpp
class Solution {
public:
    // brute force: O(N)time & space, 52.28%, ok 
    void rotate(vector<int>& nums, int k) {
        vector<int> ret;
        // great thought! If k > nums.size()
        int start = nums.size() - (k % nums.size());  
        for(int i = start; i < nums.size(); i++)
            ret.push_back(nums[i]);
        for(int i = 0; i < start; i++)
            ret.push_back(nums[i]);
        nums = ret;
    }
    
    // in place: O(kN)time, O(1)space, 0.48% 
    void rotate(vector<int>& nums, int k) {
        int size = nums.size(), it = 0;
        int iteration = k % size;
        while(it++ < iteration){
            int prev = nums[size - 1];
            for(int i = 0; i < size; i++){
                int temp = nums[i];
                nums[i] = prev;
                prev = temp;
            }
        }
    }
    
    // in place: O(N)time, O(1)space, 37.98%, ok
    void rotate(vector<int>& nums, int k) {
        int offset = k % nums.size();
        int count = 0, i = 0, prev = nums[0], start = 0;
        while(count < nums.size()){
            i = (i + offset) % nums.size();
            int temp = nums[i];
            nums[i] = prev;
            prev = temp;
            count++;
            // bug: when k = 2, it has a circle, should break it into two chains
            if(i == start) {
                i++;
                prev = nums[i];
                start = i;   // crucial!!!
            }
        }
    }
    
    // reverse three times: O(N)time, O(1)space, 52.28%, ok 
    void rotate(vector<int>& nums, int k) {
        reverse(nums.begin(), nums.end());
        reverse(nums.begin(), nums.begin() + k % nums.size());
        reverse(nums.begin() + k % nums.size(), nums.end());
    }
    
};

File: 189. Rotate Array.py
class Solution(object):
    
    # copy: O(N) time, O(N) space
    def rotate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        k %= len(nums)     # bug: k may be larger than length of nums
        temp = nums[-k:] + nums[:-k]
        for i in xrange(len(nums)):
            nums[i] = temp[i]
    
    
    # simulation: O(kN) time, O(1) space, TLE
    def rotate(self, nums, k):
        k %= len(nums)     # bug: k may be larger than length of nums
        while k:
            temp = nums[-1]
            k -= 1
            for i in xrange(len(nums) - 1, 0, -1):
                nums[i] = nums[i - 1]
            nums[0] = temp
    
    
    # replace in order: O(N) time, O(1) space
    def rotate(self, nums, k):
        if len(nums) == 0:
            return 
        currIdx, prev, visited = 0, nums[0], set()

        # starting from first elem, shift in order
        while len(visited) < len(nums):
            if currIdx in visited:
                currIdx += 1
                prev = nums[currIdx]
                
            visited.add(currIdx)
            idx = (currIdx + k) % len(nums)
            temp = nums[idx]
            nums[idx] = prev
            currIdx, prev = idx, temp

    # replace in order: O(N) time, O(1) space
    def rotate(self, nums, k):
        if len(nums) == 0:
            return 
        k %= len(nums)     # bug: k may be larger than length of nums
        nums[:-k] = reversed(nums[:-k])
        nums[-k:] = reversed(nums[-k:])
        nums[:] = reversed(nums)
                            
        
        